data-model/src/main/scala/za/co/absa/enceladus/model/menas/audit/Auditable.scala

```diff
@@ -79,6 +79,74 @@ trait Auditable[T <: Product] { self: T =>
     }).getOrElse(Seq())
   }
 
+
+
+  private[model] def getOptionalMapFieldsAudit(newValue: T, fieldName: AuditFieldName)(implicit ct: ClassTag[T]): Seq[AuditTrailChange] = {
+    def tupleToSomeString(v: (String, _)): Option[String] = Some(s"""key: "${v._1}" value: "${v._2}"""")
+
+    val index = getFieldIndex(fieldName.declaredField)
+
+    index.map({i =>
+      val maybeNewMap = newValue.productElement(i).asInstanceOf[Option[Map[String, _]]]
+      val maybeOldMap = self.productElement(i).asInstanceOf[Option[Map[String, _]]]
+
+      (maybeNewMap, maybeOldMap) match {
```

Comment:
**Suggestion:** This function is rather long. I would put the cases into sub-functions.
* Shorter
* `oldMap.toSet` and `newMap.toSet` wouldn't need to be done multiple times
* The second and the third `case` is code duplication for parts of the first `case`

Reply:
Agreed on being a bit long.

But I have the feeling that maybe only the first case processing is actually needed (because what is the auditing difference between having `None` vs `Some(Map.empty)`), so this may be a way to shorten it. I hope I am not missing something but would do:

```scala
index.map({ i =>
      // because from AuditTrail perspective None vs Some(Map.empty) yields the same AuditTrail result:
      val newMap = newValue.productElement(i).asInstanceOf[Option[Map[String, _]]].getOrElse(Map.empty)
      val oldMap = self.productElement(i).asInstanceOf[Option[Map[String, _]]].getOrElse(Map.empty)

      val added = ... // same as before
      val removed = ... // same as before
     ...

      val addedList = ... // same as before
      val removedList = ... // same as before
      val changedList = ... // same as before

      removedList ++ addedList ++ changedList
    }).getOrElse(Set.empty).toSeq

```

I believe that after this, the result might already be simple enough. But some sub-functions, as @benedeki suggested, will still be possible, of course. Your call.